ironfist/src/cpp/game/save.cpp

```diff
@@ -228,28 +247,53 @@ ironfist_map::hero_t WriteHeroXML(hero* hro) {
 	return hx;
 }
 
-void SaveMapVariables(ironfist_map::map_t& m) {
-	int key = -1;
-	const char* mapVariableId;
-	const char* mapVariableValue;
-	GetNextMapVariable(key, mapVariableId, mapVariableValue);
-	while (mapVariableId != "noMapVariables") {
-		if (mapVariableValue == NULL) {
-			char *s1 = "MapVariable '";
-			char *s2 = "' could not be saved.";
-			int len = strlen(mapVariableId) + strlen(s1) + strlen(s2) + 1;
-			char *errorMessage = (char *)ALLOC(len);
-			_snprintf_s(errorMessage, len * sizeof(*errorMessage), len, "%s%s%s", s1, mapVariableId, s2);
-			DisplayError((const char*) errorMessage, "mapVariable Error");
-			FREE(errorMessage);
-		}
-		else {
-			ironfist_map::mapVariable_t A;
-			A.id(mapVariableId);
-			A.value(mapVariableValue);
-			m.mapVariable().push_back(A);
+std::string MapVarTypeToString(MapVarType type) {
+	if (type == MAPVAR_TYPE_STRING) {
+		return "string";
+	} else if (type == MAPVAR_TYPE_NUMBER) {
+		return "number";
+	} else if (type == MAPVAR_TYPE_BOOLEAN) {
+		return "boolean";
+	} else if (type == MAPVAR_TYPE_TABLE) {
+		return "table";
+	}
+}
+
+ironfist_map::table_t WriteMapVariableTableXML(const std::string &id, luaTable lt) {
+	std::pair<luaTableElements, luaSubTables> lte_lst = lt[id];
+	ironfist_map::table_t xsdTable;
+	xsdTable.tableId(id);
+
+	for (luaTableElements::const_iterator it = lte_lst.first.begin(); it != lte_lst.first.end(); ++it) {
+		ironfist_map::tableElement_t element;
+		element.key(it->first);
+		element.type(MapVarTypeToString(it->second.first));
+		element.value(it->second.second);
+		xsdTable.tableElement().push_back(element);
+	}
+
+	for (luaSubTables::const_iterator it = lte_lst.second.begin(); it != lte_lst.second.end(); ++it) {
+		xsdTable.subTable().push_back(WriteMapVariableTableXML(*it, lt));
+	}
+	return xsdTable;
+}
+
+void WriteMapVariablesXML(ironfist_map::map_t& m) {
+
+	std::map<std::string, mapVariable> mapVariables = LoadMapVariablesFromLUA();

```

Comment:
The mapvariables are build inside that function, and contain pointers to strings and luatables. I think the data gets destructed when that function exits.